test/types/runtime.go

```diff
@@ -16,8 +16,102 @@ package types
 import (
 	"net/http"
 	"time"
+
+	"github.com/knative/pkg/ptr"
 )
 
+// MustEnvVars defines environment variables that "MUST" be set.
+// The value provided is an example value.
+var MustEnvVars = map[string]string{
+	"PORT": "8801",
+}
+
+// ShouldEnvVars defines environment variables that "SHOULD" be set.
+// To match these values with test service parameters,
+// map values must represent corresponding test.ResourceNames fields
+var ShouldEnvVars = map[string]string{
+	"K_SERVICE":       "Service",
+	"K_CONFIGURATION": "Config",
+	"K_REVISION":      "Revision",
+}
+
+// MustFiles specifies the file paths and expected permissions that MUST be set as specified in the runtime contract.
+// See https://golang.org/pkg/os/#FileMode for "Mode" string meaning. '*' indicates no specification.
+var MustFiles = map[string]FileInfo{
```

Comment:
Are some (many) of these specified by OCI?

I'm also wondering about the required permissions of /tmp and /var/log -- would it make more sense to test writing a file to those locations, rather than reading unix permissions (which may not be sufficient to grant access in some cases)?

Reply:
(Just meant here that a comment about where these are specified in OCI would help subsequent readers.)

Reply:
The first set of files are the "Default Devices" specified [here](https://github.com/knative/serving/blob/master/docs/runtime-contract.md#devices). 

The second set of files are the ["Dev symbolic links"](https://github.com/knative/serving/blob/master/docs/runtime-contract.md#dev-symbolic-links) which we had as "As specified by OCI."

In my [PR](https://github.com/knative/serving/pull/4035) to update the runtime contract I pull in the "Dev symbolic links" inline and make the "Default Devices" requirement more clear.

Reply:
As for reading of files and running these tests with and without root user containers I think these are enhancements we definitely can make. I think those tests can be layers upon this rather than replacing this.

Reply:
Created https://github.com/knative/serving/issues/4083